src/redisearch_rs/result_processor/src/lib.rs

```diff
@@ -0,0 +1,357 @@
+    /// to a `*mut Header` (which is what the C code expects to receive).
+    ///
+    /// Duplicates [`ffi::ResultProcessor`] to add pinning-related safety features.
+    ///
+    /// # Safety
+    ///
+    /// Result processors intrusively linked, where one result processor has the pointer to the
+    /// previous (forming an intrusively singly-linked list). This places a big safety invariant on all code
+    /// that touches this data structure: Result processors must *never* be moved while part of a QueryIterator
+    /// chain. Accidentally moving a processor will remove in a broken link and undefined behaviour.
+    ///
+    /// This invariant is implicit in the C code where its uncommon to move the heap allocated objects (which would
+    /// mean memcopying them around); In Rust we need to explicitly manage this invariant however, as move semantics make
+    /// it easy to break accidentally. For details refer to the [`Pin`] documentation which explains the concept of "pinning"
+    /// a Rust value in memory and its implications.
+    ///
+    /// This crate has to ensure that these pinning invariants are upheld internally, but when sending pointers through the FFI boundary
+    /// this is impossible to guarantee. Thankfully, as mentioned above its quite rare to move out of pointers, so its reasonably safe.
+    #[repr(C)]
+    #[derive(Debug)]
+    struct Header {
+        // Reference to the parent QueryIterator that owns this result processor
+        parent: *mut ffi::QueryIterator,
+        // Previous result processor in the chain
+        upstream: *mut Header,
+        // Type of result processor
+        ty: ffi::ResultProcessorType,
+        gil_time: timespec,
+        // "VTable" function. Pulls [`ffi::SearchResult`]s out of this result processor.
+
+        // Populates the result pointed to by `res`. The existing data of `res` is
+        // not read, so it is the responsibility of the caller to ensure that there
+        // are no refcount leaks in the structure.
+
+        // Users can use [`ffi::SearchResult_Clear`] to reset the structure without freeing it.
+
+        // The populated structure (if [`ffi::RPStatus_RS_RESULT_OK`] is returned) does contain references
+        // to document data. Callers *MUST* ensure they are eventually freed.
+        next: Option<unsafe extern "C" fn(self_: *mut Header, res: *mut ffi::SearchResult) -> c_int>,
+        // "VTable" function. Frees the processor and any internal data related to it.
+        free: Option<unsafe extern "C" fn(self_: *mut Header)>,
+        // ResultProcessor *must* be !Unpin to ensure they can never be moved, and they never receive
+        // LLVM `noalias` annotations; See <https://github.com/rust-lang/rust/issues/63818>.
+        // FIXME: Remove once <https://github.com/rust-lang/rust/issues/63818> is closed and replace with the recommended fix.
+        _unpin: PhantomPinned,
+    }
+}
```

Comment:
What's stopping us from moving the source of truth for this type from C to Rust? 
As long as we don't change its definition, we should be in a good place.

The `_unpin` field will just disappear when `cbindgen` does its work.

Reply:
Nothing, but IMO we should do this in a follow up PR once the Counter RP and associated FFI machinery is merged too.